clang/tools/libclang/CCAS.cpp

```diff
@@ -415,26 +416,34 @@ CXCASReplayResult clang_experimental_cas_replayCompilation(
   auto *DiagsPrinter = new TextDiagnosticPrinter(DiagOS, DiagOpts.get());
   DiagnosticsEngine Diags(DiagID, DiagOpts.get(), DiagsPrinter);
 
+  SmallVector<const char *, 256> Args(argv, argv + argc);
+  llvm::BumpPtrAllocator Alloc;
+  bool CLMode = driver::IsClangCL(
+      driver::getDriverMode(argv[0], ArrayRef(Args).slice(1)));
+
+  if (llvm::Error E = driver::expandResponseFiles(Args, CLMode, Alloc)) {
+    Diags.Report(diag::err_drv_expand_response_file)
+        << llvm::toString(std::move(E));
+    if (OutError)
+      *OutError = cxerror::create(DiagOS.str());
+    return nullptr;
+  }
+
   auto Invok = std::make_shared<CompilerInvocation>();
-  ArrayRef Argv(argv, argc);
-  bool Success = CompilerInvocation::CreateFromArgs(*Invok, Argv.drop_front(),
-                                                    Diags, Argv.front());
+  bool Success = CompilerInvocation::CreateFromArgs(*Invok, ArrayRef(Args).drop_front(),
+                                                    Diags, Args.front());
   if (!Success) {
     if (OutError)
       *OutError = cxerror::create(DiagOS.str());
     return nullptr;
   }
 
-  StringRef WorkingDir = WorkingDirectory;
-  if (!WorkingDir.empty())
-    Invok->getFileSystemOpts().WorkingDir = WorkingDir;
```

Comment:
Would we get the same crash if the caller passed the `-working-directory` clang argument?

Reply:
Not if the invocation being used for replay actually has the right cache key. In that case, the include-tree should have the absolute paths since they would be resolved by working-directory or be absolute to start with. As long as we are relying on the invocation to construct the filesystem being used here we need the invocation to match the cache key or we could get this crash. If we fixed the FIXME I added so it gets the filesystem from the cache key itself I think we'd be more robust.

Reply:
This turned out to be wrong: I thought we would have absolute paths in the include-tree when using -working-directory, but that's not the case.  More importantly, include-tree is just not working with -working-directory right now even when compiling instead of replaying.